synapse/http/replicationagent.py

```diff
@@ -0,0 +1,175 @@
+# Copyright 2023 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from typing import TYPE_CHECKING, Optional
+
+from zope.interface import implementer
+
+from twisted.internet import defer
+from twisted.internet.endpoints import HostnameEndpoint, wrapClientTLS
+from twisted.internet.interfaces import IStreamClientEndpoint
+from twisted.python.failure import Failure
+from twisted.web.client import (
+    URI,
+    HTTPConnectionPool,
+    _AgentBase,
+    _DeprecatedToCurrentPolicyForHTTPS,
+)
+from twisted.web.error import SchemeNotSupported
+from twisted.web.http_headers import Headers
+from twisted.web.iweb import (
+    IAgent,
+    IAgentEndpointFactory,
+    IBodyProducer,
+    IPolicyForHTTPS,
+    IResponse,
+)
+
+from synapse.types import ISynapseReactor
+
+if TYPE_CHECKING:
+    from synapse.server import HomeServer
+
+logger = logging.getLogger(__name__)
+
+
+@implementer(IAgentEndpointFactory)
+class ReplicationEndpointFactory:
+    """Connect to a given TCP socket"""
+
+    def __init__(
+        self,
+        reactor: ISynapseReactor,
+        hs: "HomeServer",
+        context_factory: Optional[IPolicyForHTTPS],
+    ) -> None:
+        self.reactor = reactor
+        self.hs = hs
+        self.context_factory = context_factory
+
+    def endpointForURI(self, uri: URI) -> IStreamClientEndpoint:
+        """
+        This part of the factory decides what kind of endpoint is being connected to.
+
+        Args:
+            uri: The pre-parsed URI object containing all the uri data
+
+        Returns: The correct client endpoint object
+        """
+        if b"http" in uri.scheme:
+            endpoint = HostnameEndpoint(self.reactor, uri.host, int(uri.port))
+            if uri.scheme == b"https":
+                endpoint = wrapClientTLS(self.context_factory, endpoint)
+            return endpoint
+        else:
+            raise SchemeNotSupported()
```

Comment:
I'm not sure I raised `SchemeNotSupported()` right, `uri.scheme` is the data.

Reply:
I think we just want to add an error message and this seems OK? See [the default implementation](https://github.com/twisted/twisted/blob/7e57b24a9fb20c590eac97d7b8ba0730c3633873/src/twisted/web/client.py#L1008):

```suggestion
            raise SchemeNotSupported(f"Unsupported scheme: {uri.scheme!r}")
```

Reply:
That feels way to simple, but I'm on board with that.